File: pkg/gopdflib/fill.go
// Package gopdflib provides PDF form filling functionality.
package gopdflib

import (
	"github.com/chinmay-sawant/gopdfsuit/v7/internal/pdf/form"
)

// FillPDFWithXFDF fills a PDF form with data from an XFDF file.
// XFDF (XML Forms Data Format) is an XML-based format for representing
// form data and annotations in PDF documents.
//
// WASM: the browser binding goFillPDF passes pdfBytes and xfdfBytes as two
// Uint8Array values (via js.CopyBytesToGo); this signature already takes
// plain byte slices so no adapter is needed on the Go side.
//
// Limit: the /NeedAppearances true flag is applied as a byte-level patch to
// the AcroForm dictionary. When the AcroForm dictionary lives inside a
// compressed object stream (/ObjStm) the byte patch cannot reach it, so
// viewers may not regenerate field appearances on open for such files.
// Field values written into object streams are still updated by the
// object-stream-aware fill path; only the appearance flag is affected.
//
// Example:
//
//	pdfBytes, _ := os.ReadFile("form.pdf")
//	xfdfBytes, _ := os.ReadFile("data.xfdf")
//	filled, err := gopdflib.FillPDFWithXFDF(pdfBytes, xfdfBytes)
//	if err != nil {
//	    log.Fatal(err)
//	}
//	os.WriteFile("filled.pdf", filled, 0644)
func FillPDFWithXFDF(pdfBytes, xfdfBytes []byte) ([]byte, error) {
	const op = "gopdflib: FillPDFWithXFDF"
	if len(pdfBytes) == 0 {
		return nil, invalidInputError(op, "needs a non-empty PDF")
	}
	if len(xfdfBytes) == 0 {
		return nil, invalidInputError(op, "needs non-empty XFDF data")
	}
	out, err := form.FillPDFWithXFDF(pdfBytes, xfdfBytes)
	if err != nil {
		return nil, wrapEngineError(op, err)
	}
	return out, nil
}

File: bindings/python/pypdfsuit/fill.py
"""
PDF form filling functionality.
"""

from ._bindings import get_lib, call_bytes_result, pdf_args


def fill_pdf_with_xfdf(pdf_data: bytes, xfdf_data: bytes) -> bytes:
    """
    Fill a PDF form with data from XFDF (XML Forms Data Format).

    XFDF is an XML-based format for representing form data and annotations
    in PDF documents.

    Args:
        pdf_data: The PDF form content as bytes
        xfdf_data: The XFDF data as bytes

    Returns:
        bytes: The filled PDF content

    Raises:
        GoPDFSuitError: If form filling fails
        ValueError: If pdf_data or xfdf_data is empty

    Example:
        >>> from pypdfsuit import fill_pdf_with_xfdf
        >>> with open("form.pdf", "rb") as f:
        ...     pdf_data = f.read()
        >>> with open("data.xfdf", "rb") as f:
        ...     xfdf_data = f.read()
        >>> filled = fill_pdf_with_xfdf(pdf_data, xfdf_data)
        >>> with open("filled.pdf", "wb") as f:
        ...     f.write(filled)
    """
    lib = get_lib()
    return call_bytes_result(lib.FillPDFWithXFDF, *pdf_args(pdf_data), *pdf_args(xfdf_data, "XFDF data"))
